Project Autocomplete Controller

reynoldtan opened the pull request:
<!--- Thank you for contributing! -->
<!--- Provide a general summary of your changes in the Title above -->
<!--- See our Contribution Guidelines here:
          https://github.com/tripal/tripal/blob/7.x-3.x/CONTRIBUTING.md -->


<!---  Please set the header below based on the PR type:
# New Feature
# Bug Fix
# Tripal 4 Core Dev Task  --->

#

Issue https://github.com/tripal/tripal/issues/1553

<!--- Enter the Tripal version this PR applies to (i.e. either 3 or 4 ;-p) --->
### Tripal Version: 

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
This is a CHADO-focused autocomplete for projects. Check the issue for more motivation for this PR.

This PR will create a route that can be used as a value to the #autocomplete_route_name when creating an autocomplete form element. It includes a controller which follows the template provided by Tripal Term Autocomplete and queries CHADO using Tripal DBX to suggest projects based on user input.

## Testing?
<!--- Please describe in detail how to test these changes. -->
<!--- Reviewers will use this section to test the submission! -->
<!--- If you've implemented PHPUnit tests, you can describe the test cases here. -->
This PR includes automated tests which will test:
1. Insert test projects.
2. The class handler to evaluate number of suggestions returned and if desired project was part of the suggestions.
3. Support classes getProjectName and getProjectId to resolve project name (given a project id) and project id (given a project name), respectively.
4. Invalid inputs to all methods.

Manual Testing.
There is no place in Tripal Core that needs this autocomplete. An example form element would look this:

```
 $form['my_autocomplete'] = [
   '#type' => 'textfield',
   '#autocomplete_route_name' => 'tripal_chado.project_autocomplete',
   '#autocomplete_route_parameters' => ['type_id' => 0, 'count' => 5]
 ];

// Where type_id is the type id number used to create types in projectprop table to restrict results to only match this id. 
// In this example it is set to 0 to return all projects regardless of type.
```

Since the autocomplete returns human-readable output (project name) and not the project id, there is an additional method that will fetch the project id of a given project name.

```
 $project_id = ChadoProjectAutocompleteController::getProjectId($project_name);
```

To set default value (#default_value property) to fields, where stored value is the project_id, a method to resolve project id to project name is provided by the method getProjectName().

```
$project_name = ChadoProjectAutocompleteController::getProjectName($project_id);
```

reynoldtan committed:
Project Autocomplete Controller

laceysanderson committed:
Merge branch '4.x' into tv4g2-issue1553-chadoProjectAutocomplete

dsenalik committed:
Merge branch '4.x' into tv4g2-issue1553-chadoProjectAutocomplete

dsenalik committed:
Merge branch '4.x' into tv4g2-issue1553-chadoProjectAutocomplete

dsenalik commented:
I made a docker with this branch, and set up testing as follows.
1. Create some test projects, I created these 
![Test Projects](https://github.com/tripal/tripal/assets/8419404/17262039-5c7f-40e6-84da-5769a9de610b)
2. Create the projectprop table entries, I picked some random cvterm_id values, they are not significant (I think)
```
insert into chado.projectprop (project_id, type_id) values (1, 10), (2, 10), (3, 10), (4, 11);
INSERT 0 4
sitedb=> select * from chado.projectprop;
 projectprop_id | project_id | type_id | value | rank 
----------------+------------+---------+-------+------
              1 |          1 |      10 |       |    0
              2 |          2 |      10 |       |    0
              3 |          3 |      10 |       |    0
              4 |          4 |      11 |       |    0
(4 rows)
```
3. Put the form field somewhere to test. I added it to `tripal/src/Form/TripalImporterForm.php` because that's easy to get to: Tripal → Data Loaders → OBO Vocabulary Loader
At line 47 I inserted this
```
$form['my_autocomplete'] = [
   '#type' => 'textfield',
   '#autocomplete_route_name' => 'tripal_chado.project_autocomplete',
   '#autocomplete_route_parameters' => ['type_id' => 10, 'count' => 5],
   '#title' => 'Test autocomplete title',
   '#description' => 'Test autocomplete description',
];
```

Unfortunately when I test this with a single letter, nothing happens, it sticks at Loading... and nothing happens
![StuckRunning](https://github.com/tripal/tripal/assets/8419404/6077738c-e070-4c5e-ac71-ee1e46cf2f2d)

There are three errors in the log

Warning: Undefined array key "function" in Drupal\Core\Utility\Error::getLastCaller() (line 150 of /var/www/drupal9/web/core/lib/Drupal/Core/Utility/Error.php) 

Symfony\Component\HttpKernel\Exception\ControllerDoesNotReturnResponseException: The controller must return a "Symfony\Component\HttpFoundation\Response" object but it returned null. Did you forget to add a return statement somewhere in your controller? in () (line 98 of /var/www/drupal9/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php).

Warning: Undefined array key "function" in Drupal\Core\Utility\Error::getLastCaller() (line 150 of /var/www/drupal9/web/core/lib/Drupal/Core/Utility/Error.php) 

Only when you enter a second letter does the autocomplete kick in (this is by design in the code to wait until two letters are entered). 

If I comment out at line 75 in `tripal_chado/src/Controller/ChadoProjectAutocompleteController.php`
```
//    if ($response) {
      return new JsonResponse($response);
//    }
```
then there are no errors generated in the log.

![AB](https://github.com/tripal/tripal/assets/8419404/0053875d-b865-4259-b530-713e5670c043)

And if I enter "BC" I get only one project "ABC", as expected, not the "BCD" project with a different type_id. This is the correct desired behavior.

An issue, though, is the annoying spinning circle if you only enter 1 letter, or if there is no match to the letters you entered, if I enter "XYZ" for example. But I'm not sure how that normally works...

dsenalik commented:
I figured it out, in addition to the change above, in `tripal_chado/src/Controller/ChadoProjectAutocompleteController.php` line 32 change
```
    // Array to hold matching project records.
    $response = null;
```
to
```
    // Array to hold matching project records.
    $response = [];
```

dsenalik commented on tripal_chado/src/Controller/ChadoProjectAutocompleteController.php:
```diff
@@ -0,0 +1,133 @@
+   */
+  public function handleAutocomplete(Request $request, int $type_id = 0, int $count = 5) {
+    // Array to hold matching project records.
+    $response = null;
```
change to $response = [];

dsenalik commented on tripal_chado/src/Controller/ChadoProjectAutocompleteController.php:
```diff
@@ -0,0 +1,133 @@
+      }  
+    }
+    
+    if ($response) {
+      return new JsonResponse($response);
+    }
```
remove the `if ($response)`, always want to return a response
`return new JsonResponse($response);`

dsenalik commented:
@reynoldtan can you please make the same change to initialize `$response = [];` in `tripal_chado/src/Controller/ChadoCVTermAutocompleteController.php` line 30. It seems like a lot of work to create a whole separate issue for such a minor change, I feel we can add it in here since it is the exact same change.
Thanks

Actually we may instead fix it in pull #1488, see the comment here https://github.com/tripal/tripal/pull/1488#pullrequestreview-1509624678 the first two points are relevant to the CV autocomplete.

reynoldtan committed:
Initialize response variable to empty array and begin autocomplete at 1 character entered

reynoldtan committed:
Merge branch 'tv4g2-issue1553-chadoProjectAutocomplete' of https://github.com/reynoldtan/tripal into tv4g2-issue1553-chadoProjectAutocomplete

reynoldtan commented:
Hi everyone, thanks for the review and as recommended, I have revised autocomplete handler with the correct initial value for the $response variable and removed the condition of the return value. I have also removed the requirement to begin autocomplete when 2 characters have been entered.

dsenalik reviewed (approved):
Thanks

laceysanderson reviewed (approved):
Thanks to @dsenalik thorough functional review, I just did a code review on this one.

✅  code review looks good

Thanks @reynoldtan ❤️ 🎉 
This one is ready to merge too

laceysanderson committed:
Merge branch '4.x' into tv4g2-issue1553-chadoProjectAutocomplete
